build scripts: various cleanups and minor changes preparing for containerized build and test

phlogistonjohn opened the pull request:
The first patch is directly from @epuertat original patch, but just the shell script parts.

The remainder is a bunch of small things that I found while working on running the build and test scripts within a centos8 container. Some of them fix actual blockers to running in the container. Others are there to try and slightly clean up, reorganise, and de-duplicate parts of the scripts.

In particular, I'm adding a new src/script/lib-build.sh to share some common bits and parts of the current scripts pulled into new functions.  Beyond getting some things to work in the centos 8 container environment there should be no externally visible behavior changes to the existing scripts.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

adk3798 commented:
jenkins test make check

adk3798 commented:
last make check failure
```
FAILED: src/rgw/CMakeFiles/rgw_common.dir/rgw_sal.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAVE_CONFIG_H -DM0_EXTERN=extern -DM0_INTERNAL="" -D_ARROW_EXIST -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/services -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw -I/usr/include/lua5.3 -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/dmclock/src -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/dmclock/support/src -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/spawn/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/fmt/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/store/dbstore -isystem /opt/ceph/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/rapidjson/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/src/arrow/include -isystem /usr/include/motr -g -Werror -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -Wno-attributes -std=c++20 -MD -MT src/rgw/CMakeFiles/rgw_common.dir/rgw_sal.cc.o -MF src/rgw/CMakeFiles/rgw_common.dir/rgw_sal.cc.o.d -o src/rgw/CMakeFiles/rgw_common.dir/rgw_sal.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_sal.cc
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_sal.cc:36:
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_sal_motr.h:25:
In file included from /usr/include/motr/motr/client.h:1875:
In file included from /usr/include/motr/cas/cas.h:40:
In file included from /usr/include/motr/fop/fom_interpose.h:28:
In file included from /usr/include/motr/fop/fom.h:212:
In file included from /usr/include/motr/stob/stob.h:38:
/usr/include/motr/be/extmap.h:112:10: fatal error: 'btree/btree.h' file not found
#include "btree/btree.h"   /* m0_btree_cursor */
         ^~~~~~~~~~~~~~~
```

phlogistonjohn commented:
Same failure. I didn't think my compiler selection patches should have an externally visible result... but now I'm not so sure. I will just restart the job a couple more times but if it keeps failing this way, I'll drop the compiler related patches from the branch and see if that makes a difference.

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
Different this time:
```
In file included from /home/jenkins-build/build/workspace/ceph-api/src/rgw/rgw_sal.cc:36:
In file included from /home/jenkins-build/build/workspace/ceph-api/src/rgw/rgw_sal_motr.h:25:
In file included from /usr/include/motr/motr/client.h:1875:
/usr/include/motr/cas/cas.h:39:10: fatal error: 'fop/wire_xc.h' file not found
#include "fop/wire_xc.h"
         ^~~~~~~~~~~~~~~
1 error generated.
```

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
Now another error (?):
```
/usr/bin/clang-14 -Wl,-z,relro -Wl,--fatal-warnings -Wl,--warn-common  -Wl,-rpath=/usr/local/lib/pmdk_debug:: -L../debug/libpmem2/.. -shared -Wl,--version-script=libpmem2.link,-soname,libpmem2.so.1 -o ../debug/libpmem2/../libpmem2.so.1.0.0 ../debug/libpmem2/ravl.o ../debug/libpmem2/libpmem2.o ../debug/libpmem2/badblocks.o ../debug/libpmem2/badblocks_none.o ../debug/libpmem2/config.o ../debug/libpmem2/deep_flush.o ../debug/libpmem2/errormsg.o ../debug/libpmem2/memops_generic.o ../debug/libpmem2/map.o ../debug/libpmem2/map_posix.o ../debug/libpmem2/persist.o ../debug/libpmem2/persist_posix.o ../debug/libpmem2/pmem2_utils.o ../debug/libpmem2/usc_none.o ../debug/libpmem2/source.o ../debug/libpmem2/source_posix.o ../debug/libpmem2/vm_reservation.o ../debug/libpmem2/vm_reservation_posix.o ../debug/libpmem2/ravl_interval.o ../debug/libpmem2/auto_flush_linux.o ../debug/libpmem2/deep_flush_linux.o ../debug/libpmem2/extent_linux.o ../debug/libpmem2/pmem2_utils_linux.o ../debug/libpmem2/pmem2_utils_none.o ../debug/libpmem2/region_namespace_none.o ../debug/libpmem2/alloc.o ../debug/libpmem2/fs_posix.o ../debug/libpmem2/os_posix.o ../debug/libpmem2/os_thread_posix.o ../debug/libpmem2/out.o ../debug/libpmem2/util.o ../debug/libpmem2/util_posix.o ../debug/libpmem2/init.o ../debug/libpmem2/cpu.o ../debug/libpmem2/memcpy_nt_avx.o ../debug/libpmem2/memcpy_nt_sse2.o ../debug/libpmem2/memset_nt_avx.o ../debug/libpmem2/memset_nt_sse2.o ../debug/libpmem2/memcpy_t_avx.o ../debug/libpmem2/memcpy_t_sse2.o ../debug/libpmem2/memset_t_avx.o ../debug/libpmem2/memset_t_sse2.o ../debug/libpmem2/memcpy_nt_avx512f.o ../debug/libpmem2/memset_nt_avx512f.o ../debug/libpmem2/memcpy_t_avx512f.o ../debug/libpmem2/memset_t_avx512f.o    -pthread 
ln -sf libpmem2.so.1.0.0 ../debug/libpmem2/../libpmem2.so.1
ln -sf libpmem2.so.1 ../debug/libpmem2/../libpmem2.so
gmake[2]: Leaving directory '/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/libpmem2'
/usr/bin/gmake -C librpmem
gmake[2]: Entering directory '/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/librpmem'
NOTE: Skipping librpmem because libfabric (version >= 1.4.2) is missing -- see src/librpmem/README for details
mkdir -p ../nondebug/librpmem
gmake[2]: Leaving directory '/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/librpmem'
/usr/bin/gmake -C librpmem DEBUG=1
gmake[2]: Entering directory '/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/librpmem'
NOTE: Skipping librpmem because libfabric (version >= 1.4.2) is missing -- see src/librpmem/README for details
mkdir -p ../debug/librpmem
gmake[2]: Leaving directory '/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/librpmem'
gmake[1]: Leaving directory '/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src'
[1851/2618] Performing build step for 'spdk-ext'
ninja: build stopped: subcommand failed.
Build step 'Execute shell' marked build as failure
[Cobertura] Skipping Cobertura coverage report as build was not SUCCESS or better ...
```

PS. This was actually the very first failure case which I hadn't captured in a comment.

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
```
In file included from /usr/include/motr/motr/client.h:1875:
/usr/include/motr/cas/cas.h:38:10: fatal error: 'fop/wire.h' file not found
#include "fop/wire.h"           /* m0_fop_mod_rep */
```
Again

phlogistonjohn commented:
I've dropped 2 compiler detection related patches from the series, to see if there's any impact on the failures of `make check`.

phlogistonjohn commented:
First run after dropping the two compiler related patches worked. Will add them back one at a time now (half-baked bisect like).

phlogistonjohn commented:
Hmm. This time it failed but got past the compile. It died at unittest_osdmap.

epuertat commented on src/script/run-make.sh:
```diff
@@ -58,76 +48,26 @@ function detect_ceph_dev_pkgs() {
-        echo "This probably means distribution $ID is not supported by run-make-check.sh" >&2
+    if test -f ./install-deps.sh ; then
+        ci_debug "Running install-deps.sh"
+        INSTALL_EXTRA_PACKAGES="ccache git $which_pkg clang"
```
Nice! I racked my brains trying to remove package installs from here!

epuertat commented on install-deps.sh:
```diff
@@ -545,4 +568,4 @@ if $for_make_check; then
     type git > /dev/null || (echo "Dashboard uses git to pull dependencies." ; false)
```
And we should probably rewrite this... Dashboard used `git` long time ago since some NPM dependency was pinned to a github repo, instead to an npmjs published package, but not anymore. However, `git` is used by the `run-make-check` script to obtain the SHA of the current branch.

phlogistonjohn replied:
OK, good to know.

epuertat commented on run-make-check.sh:
```diff
@@ -37,10 +33,14 @@ function run() {
+    if [ "${procs}" -gt $m ]; then
+        m="${procs}"
+    fi
+    local aiomax="$((65536 * procs))"
+    if [ "$(/sbin/sysctl -n fs.aio-max-nr )" -lt "${aiomax}" ]; then
+        $DRY_RUN sudo /sbin/sysctl -q -w fs.aio-max-nr="${aiomax}"
```
Once we move to containers, we should remove this... I'd be worried if some building script would mess with my sysctl settings (and that's only for the testing stage). In fact, we should be able to run these scripts without sudo.

phlogistonjohn replied:
Yep. One thing I almost did for this series but skipped for now was to replace calls to `sudo` to a function like `try_sudo` which checks the EUID by default and can be forced on/off by a variable. Again, to keep the series smaller and easier to review I defered that until later. As for this particular line, I agree that it's not good to do inside the container... but I left it for now.

epuertat commented on src/script/lib-build.sh:
```diff
@@ -0,0 +1,88 @@
+    local cxx_compiler=g++
+    local c_compiler=gcc
+    # ubuntu/debian ci builds prefer clang
+    for i in $(seq 14 -1 10); do
```
What about?
```suggestion
    for i in {14..10}; do
```

phlogistonjohn replied:
Sure.

epuertat reviewed (approved):
Great work, @phlogistonjohn! I just left some suggestions, think-aloud ideas, ...

phlogistonjohn commented on src/script/lib-build.sh:
```diff
@@ -0,0 +1,88 @@
+# The following global only exists to help detect if lib-build has already been
+# sourced. This is only needed because the scripts that are being migrated are
+# often sourcing (as opposed to exec'ing one another).
+# shellcheck disable=SC2034
```
OK, I'll think about it. I usually try to make smaller patches and PRs that only do one thing at a time, but if its not a huge effort, and I can borrow from something existing, I can add it to this pr too.

phlogistonjohn commented on src/script/run-make.sh:
```diff
@@ -148,24 +89,61 @@ EOM
+        ci_debug "Failed to discover a compiler"
+    fi
+    if [ "${discovered_compiler_env}" ]; then
+        ci_debug "I should not see this on ubutnu (2)"
```
I've already removed that log line and replaced it with something else, thanks! :-)

phlogistonjohn commented on src/script/lib-build.sh:
```diff
@@ -0,0 +1,88 @@
+# `discovered_compiler_env`. The `discovered_compiler_env` variable
+# may be blank. If not, it will contain a file that needs to be sourced
+# prior to using the compiler.
+function discover_compiler() {
```
Ideally yes, but: this is mostly a move of existing code that I think exists to make it prefer one compiler over another. Also, in the centos 8 environment the "SCL" files must be sourced prior to running a tool like cmake anyway.

phlogistonjohn commented on install-deps.sh:
```diff
@@ -63,7 +68,7 @@ function munge_debian_control {
 
 function ensure_decent_gcc_on_ubuntu {
-    in_jenkins && echo "CI_DEBUG: Start ensure_decent_gcc_on_ubuntu() in install-deps.sh"
+    ci_debug "Start ensure_decent_gcc_on_ubuntu() in install-deps.sh"
```
Interesting, but I don't think we want it for every function. At least, that's not how it was being used before. It's an interesting thought though that we could revisit for the future.

phlogistonjohn commented on install-deps.sh:
```diff
@@ -12,6 +12,14 @@
 #  version 2.1 of the License, or (at your option) any later version.
 #
 set -e
```
Sounds reasonable, if I'll enable it if it doesn't break anything in the current jenkins or the containerized flow.

phlogistonjohn commented on install-deps.sh:
```diff
@@ -253,6 +258,67 @@ EOF
     fi
 }
 
+function populate_wheelhouse() {
```
Right. FWIW, I have ccache enabled in the container workflow, based on your original patches, but unfortunately it doesn't seem to work as desired. It's one of those things I want to look into soon.  But since I don't want to keep sitting on all the work done so far this PRs about getting some of the script improvements out sooner. :-)

phlogistonjohn commented on install-deps.sh:
```diff
@@ -12,6 +12,14 @@
+if ! [ "${_SOURCED_LIB_BUILD}" = 1 ]; then
+    SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+    CEPH_ROOT="${SCRIPT_DIR}"
+    . "${CEPH_ROOT}/src/script/lib-build.sh" || exit 2
```
I'll think about it. Since these are all *bash* scripts, I don't think we should hesitate to use bash specific features. Personally, I think some of the code could be improved by use of bash arrays, for example.

phlogistonjohn commented on src/script/lib-build.sh:
```diff
@@ -0,0 +1,87 @@
+}
+
+function ci_debug() {
+    if in_jenkins || [ "${FORCE_CI_DEBUG}" ]; then
```
done.

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
I'm unsure if I'm massively unlucky or somehow a code path that is not taken, as indicated by the logging, is somehow making the builds fail.  I'm going to gather some stats on these failures in an offline file. Expect to see a fair bit of noise on this PR for a while. Thanks for your patience.

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
jenkins test make check

epuertat commented:
> Between the original version of these patches and today the motr support for rgw was disabled. But because the code was moved prior to this august, this version of the cmake flags kept trying to enable motr and it was failing.
> During the debugging process I reduced the parallel builds to 2, and this seems to break generating header files. I suspect this is actually an issue in the cmake files but since that number of "processors" is artificially low I'll leave that underlying issue for another day.

A similar thing happened to me last time I had to rebase the original PR... Wouldn't it be nice to keep the flags detached from the scripts ("configuration separated from code")? Nowadays it's a common practice to have a separate/optional `.env` file which is just a series of "KEY=VALUE" statements (bash arrays work great for this):
```bash
cmake_opts=(
-DCMAKE_CXX_COMPILER=$cxx_compiler
-DCMAKE_C_COMPILER=$c_compiler
-DCMAKE_CXX_FLAGS_DEBUG=-Werror
-DENABLE_GIT_VERSION=OFF
...
-DWITH_RBD_MIRROR=ON
)
```
And then you just dump all those:
```bash
./do_cmake.sh ${cmake_opts[@]}
```

phlogistonjohn commented:
@tchaikov @idryomov & @epuertat,  @adk3798 
Hi all, can I get some additional feedback on this PR?  Or is it unexciting enough to be largely unobjectionable? ;-)
Thanks!

tchaikov commented:
@phlogistonjohn sorry for the latency. i was reading these commits couple weeks ago. they were nicely structured. and looked great in general. i will take a closer look and provide some inputs by the end this week.

tchaikov commented on install-deps.sh:
```diff
@@ -493,6 +500,9 @@ EOF
                 ;;
         esac
+        if [ "$INSTALL_EXTRA_PACKAGES" ]; then
+            $SUDO dnf install -y $INSTALL_EXTRA_PACKAGES
```
my concern is, what if the package names are different in debian derivatives and the rpm based distros?

phlogistonjohn replied:
In this case, the script/program setting INSTALL_EXTRA_PACKAGES will be responsible for passing the correct package names for the distro. This was the case before as well. The advantage of INSTALL_EXTRA_PACKAGES is that install-deps.sh script can order the install & manage the system in one central place.

For what it's worth, my WIP branch is working on both ubutnu and centos 8 stream at this time.

tchaikov commented on install-deps.sh:
```diff
@@ -516,6 +526,9 @@ EOF
         zypp_install="zypper --gpg-auto-import-keys --non-interactive install --no-recommends"
         $SUDO $zypp_install systemd-rpm-macros rpm-build || exit 1
+        if [ "$INSTALL_EXTRA_PACKAGES" ]; then
+            $SUDO dnf install -y $INSTALL_EXTRA_PACKAGES
```
not sure if `dnf` works for SuSE distros. probably should use `$SUDO $zypp_install` here?

phlogistonjohn replied:
Good catch, will fix.

phlogistonjohn replied:
Fixed.

tchaikov commented on install-deps.sh:
```diff
@@ -395,7 +395,14 @@ else
         if [ "$INSTALL_EXTRA_PACKAGES" ]; then
-            $SUDO apt-get install -y $INSTALL_EXTRA_PACKAGES
+            if ! $SUDO apt-get install -y $INSTALL_EXTRA_PACKAGES ; then
+                # try again. ported over from run-make.sh. Not sure why this is
```
there is chance that, for instance, `apt-get` is interrupted when installing / removing packages, in that case, the package management system would be in an inconsistent state. if we run `apt-get install ...` when the system is in this state, the tool would prompt us to run `dpkg --configure -a`.  i think the commit message of e2782954088dadb9bfe9cd3760129f06c602e7dd does explain this .

phlogistonjohn replied:
Thanks! I'll update the comment to match.

phlogistonjohn replied:
Updated.

tchaikov commented on src/script/lib-build.sh:
```diff
@@ -71,6 +71,15 @@ function discover_compiler() {
         fi
     done
+    # but if this is {centos,rhel} we need gcc-toolset
+    if [ -f "/opt/rh/gcc-toolset-11/enable" ]; then
```
now that GTS-12 is out on el9. and GCC-13 will be out early next year. i'd suggest use a loop for detecting GTS, so this script is more future proof, something like
```sh
for i in {13..11}; do
  # ...
done
```

phlogistonjohn replied:
Unfortunately, I tried something like this and there's a version of gcc-toolset-12 present in the centos 8 container that fails the "configure" step. I would like to defer this until I can test on centos 9 stream (and this is blocked on a copr repository issue).

tchaikov replied:
this is interesting. i've been building Ceph on rocky linux 9 with GTS-12 for a months. what was the failure you ran into? but this is not a blocker. i am just curious.

phlogistonjohn replied:
The failure was on centos (stream)8. I'm not sure why but it seems that there's a partial (?) install of 12 when 11 is installed. I plan on investigating it but for now only gcc-toolset-11 works on centos 8.

phlogistonjohn commented:
jenkins test make check

phlogistonjohn commented:
Hi @tchaikov, are you OK with my updates and explanations/justifications? Thanks.

phlogistonjohn commented:
jenkins test api

phlogistonjohn commented:
Hi @epuertat  @tchaikov -- I just rebased and that seems to have taken care of the failing API tests. I personally think this is a mergable state. Are there any other blockers you all think I need to resolve before it can be merged? Thanks!

tchaikov commented:
> Hi @tchaikov, are you OK with my updates and explanations/justifications? Thanks.

@phlogistonjohn hi John, i will re-review this PR in this week. thank you for your patience.

tchaikov commented on install-deps.sh:
```diff
@@ -395,7 +395,16 @@ else
         echo "Using apt-get to install dependencies"
         if [ "$INSTALL_EXTRA_PACKAGES" ]; then
-            $SUDO apt-get install -y $INSTALL_EXTRA_PACKAGES
+            if ! $SUDO apt-get install -y $INSTALL_EXTRA_PACKAGES ; then
```
i think the commit message does not hold anymore. as i tried to explain it in https://github.com/ceph/ceph/pull/48697#discussion_r1034672915 and the link referenced in that comment. if you are still not sure, in which part you are not convinced?

phlogistonjohn replied:
Oh, I did update the comment but not the commit message. Let me do that now. I wish github made it easier to comment on the commit messages.

tchaikov commented on src/script/lib-build.sh:
```diff
@@ -48,3 +48,32 @@ function get_processors() {
+    export discovered_c_compiler="${c_compiler}"
+    export discovered_cxx_compiler="${cxx_compiler}"
+    export discovered_compiler_env="${compiler_env}"
+    return 0
```
this is not necessary.

phlogistonjohn replied:
I'll revisit this later. I wanted something to _scream_ these are the variables to use, but yes it's a bit redundant.

tchaikov reviewed (approved):
apart from the commit message regarding `dpkg --configure -a`, looks good to me.

phlogistonjohn commented:
> apart from the commit message regarding `dpkg --configure -a`, looks good to me.

I fixed up the commit message on the patch that moves that code around. I've also rebased.

phlogistonjohn commented:
Hello @tchaikov @epuertat - given that I've got both your approvals and I fixed the commit message regarding "dpkg --configure -a" - can I bother one of you to begin merging the PR? Let me know of there are any other blocking tasks on my side. Thanks!

phlogistonjohn commented:
Hello @tchaikov @epuertat - I'd like to keep this PR on your radar and ask if there's no blockers left for me to change, that this be considered for merge.

phlogistonjohn commented:
Well, it looks like it's been a month or so since I last pinged on this PR. I thought it was shorter. Anyway, I just rebased for good measure.

@adk3798  if I don't get any additional attention from @epuertat or @tchaikov - because they've approved but (maybe?) don't want to be involved in merging this, is there a typical process in ceph for this situation?

tchaikov commented:
@phlogistonjohn hi John, i am sorry for keeping you waiting and waiting. this is frustrating. and i can feel what you feel. but i don't have enough bandwidth working on this project as a responsible maintainer anymore. i could merge your PR but IMHO, that implies more responsibilities.

could you ping @ceph/core for assistance on this?

phlogistonjohn commented:
@tchaikov thanks I appreciate the response.  I'll ping that group and I may also mail the devel list as well.

@ceph/core any thoughts on how to make progress on the PR. Its part of a general effort I took on (that Ernesto started) to make the build (and unit tests) work within containers to decouple the build from the host OS and make it easier to build locally.

adk3798 commented:
https://pulpito.ceph.com/adking-2023-02-13_21:57:03-orch:cephadm-wip-adk-testing-2023-02-13-1359-distro-default-smithi/

11 failures, one dead job

- 8 of the failed jobs were on new tests being added in by other PRs in the run, which obviously aren't quite working yet (test_monitoring_stack_basic, rotate-ssh-keys, test_iscsi_container) but outside of iscsi and monitoring stack specific changes and the PRs that introduced the tests, this shouldn't block merging.
- 2 instances of https://tracker.ceph.com/issues/58535
- 1 instance of https://tracker.ceph.com/issues/49287
- 1 dead job was an error re-imaging machines, infra issue


For this PR in particular, I think I'll also run a few tests that use the ubuntu packages and get back to this

adk3798 commented:
> https://pulpito.ceph.com/adking-2023-02-13_21:57:03-orch:cephadm-wip-adk-testing-2023-02-13-1359-distro-default-smithi/
> 
> 11 failures, one dead job
> 
> * 8 of the failed jobs were on new tests being added in by other PRs in the run, which obviously aren't quite working yet (test_monitoring_stack_basic, rotate-ssh-keys, test_iscsi_container) but outside of iscsi and monitoring stack specific changes and the PRs that introduced the tests, this shouldn't block merging.
> * 2 instances of https://tracker.ceph.com/issues/58535
> * 1 instance of https://tracker.ceph.com/issues/49287
> * 1 dead job was an error re-imaging machines, infra issue
> 
> For this PR in particular, I think I'll also run a few tests that use the ubuntu packages and get back to this

aforementioned set of jobs run on ubuntu packages https://pulpito.ceph.com/adking-2023-02-27_15:21:08-fs-wip-adk-testing-2023-02-20-1650-distro-default-smithi. Will see how those go. I don't know what known failures exist in that suite but I'm assuming if this really broke something it would affect all or most of the tests so as long as we're mostly green I think we're okay.

adk3798 commented:
> > https://pulpito.ceph.com/adking-2023-02-13_21:57:03-orch:cephadm-wip-adk-testing-2023-02-13-1359-distro-default-smithi/
> > 11 failures, one dead job
> > 
> > * 8 of the failed jobs were on new tests being added in by other PRs in the run, which obviously aren't quite working yet (test_monitoring_stack_basic, rotate-ssh-keys, test_iscsi_container) but outside of iscsi and monitoring stack specific changes and the PRs that introduced the tests, this shouldn't block merging.
> > * 2 instances of https://tracker.ceph.com/issues/58535
> > * 1 instance of https://tracker.ceph.com/issues/49287
> > * 1 dead job was an error re-imaging machines, infra issue
> > 
> > For this PR in particular, I think I'll also run a few tests that use the ubuntu packages and get back to this
> 
> aforementioned set of jobs run on ubuntu packages https://pulpito.ceph.com/adking-2023-02-27_15:21:08-fs-wip-adk-testing-2023-02-20-1650-distro-default-smithi. Will see how those go. I don't know what known failures exist in that suite but I'm assuming if this really broke something it would affect all or most of the tests so as long as we're mostly green I think we're okay.

14/15 passed, looks like we're doing okay there
